File: src/main/java/simplifying/method/calls/Lottery.java
package simplifying.method.calls;

import java.util.*;

public class Lottery {
    private static final Random RANDOM = new Random(42);
    private final HashMap<UUID, LotteryTicket> tickets = new HashMap<>();

    public String purchaseTicketForCustomer(UUID id) {
        String ticketNumber = generateTicketNumber();
        tickets.put(id, new LotteryTicket(ticketNumber, id));

        return ticketNumber;
    }

    public LotteryTicket drawWinner() {
        if (tickets.isEmpty()) {
            throw new IllegalStateException("No tickets");
        }

        List<LotteryTicket> randomizedTickets = new ArrayList<>(tickets.values());
        Collections.shuffle(randomizedTickets);

        return randomizedTickets.get(0);
    }

    private String generateTicketNumber() {
        return String.format("%06d", RANDOM.nextInt(1000000));
    }
}

File: src/scala/src/main/scala/simplifying/method/calls/Lottery.scala
package simplifying.method.calls

import java.util.UUID
import scala.util.Random

object Lottery {
  private val random: Random = new Random(42)
}

class Lottery {
  final private var tickets = Map.empty[UUID, LotteryTicket]

  def purchaseTicketForCustomer(id: UUID): String = {
    val ticketNumber: String = generateTicketNumber()
    tickets += (id -> LotteryTicket(ticketNumber, id))
    ticketNumber
  }

  private def generateTicketNumber(): String =
    String.format("%06d", Lottery.random.nextInt(1000000))

  def drawWinner(): Option[LotteryTicket] = {
    tickets.values match {
      case Nil => None
      case v   => Random.shuffle(v).headOption
    }
  }
}
